src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs

```diff
@@ -0,0 +1,2135 @@
+            finally
+            {
+                //if (NetEventSource.IsEnabled) NetEventSource.Enter(NetEventSource.ComponentType.HttpListener, this, "BeginGetContext", "IAsyncResult#" + LoggingHash.HashString(asyncResult));
+            }
+
+            return asyncResult;
+        }
+
+        public HttpListenerContext EndGetContext(IAsyncResult asyncResult)
+        {
+            //if (NetEventSource.IsEnabled) NetEventSource.Enter(NetEventSource.ComponentType.HttpListener, this, "EndGetContext", "IAsyncResult#" + LoggingHash.HashString(asyncResult));
+            HttpListenerContext httpContext = null;
+            try
+            {
+                CheckDisposed();
+                if (asyncResult == null)
+                {
+                    throw new ArgumentNullException("asyncResult");
+                }
+                //GlobalLog.Print("HttpListener#" + LoggingHash.HashString(this) + "::EndGetContext() asyncResult:" + LoggingHash.ObjectToString(asyncResult));
+                ListenerAsyncResult castedAsyncResult = asyncResult as ListenerAsyncResult;
+                if (castedAsyncResult == null || castedAsyncResult.AsyncObject != this)
+                {
+                    throw new ArgumentException(SR.net_io_invalidasyncresult, "asyncResult");
+                }
+                if (castedAsyncResult.EndCalled)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndGetContext"));
+                }
+                castedAsyncResult.EndCalled = true;
+                httpContext = castedAsyncResult.InternalWaitForCompletion() as HttpListenerContext;
+                if (httpContext == null)
+                {
+                    Debug.Assert(castedAsyncResult.Result is Exception, "EndGetContext|The result is neither a HttpListenerContext nor an Exception.");
+                    throw castedAsyncResult.Result as Exception;
+                }
+            }
+            //catch (Exception exception)
+            //{
+            //    if (NetEventSource.IsEnabled) NetEventSource.Exception(NetEventSource.ComponentType.HttpListener, this, "EndGetContext", exception);
+            //    throw;
+            //}
+            finally
+            {
+                //GlobalLog.Print("HttpListener#" + LoggingHash.HashString(this) + "::EndGetContext() returning HttpListenerContext#" + LoggingHash.ObjectToString(httpContext));
+                //if (NetEventSource.IsEnabled) NetEventSource.Exit(NetEventSource.ComponentType.HttpListener, this, "EndGetContext", httpContext == null ? "<no context>" : "HttpListenerContext#" + LoggingHash.HashString(httpContext) + " RequestTraceIdentifier#" + httpContext.Request.RequestTraceIdentifier);
+            }
+            return httpContext;
+        }
+
+        public Task<HttpListenerContext> GetContextAsync()
+        {
+            return Task<HttpListenerContext>.Factory.FromAsync(BeginGetContext, EndGetContext, null);
```

Comment:
This would be better as:
```C#
return Task.Factory.FromAsync<HttpListenerContext>(
    (callback, state) => ((HttpListener)state).BeginGetContext(callback, state),
    iar => ((HttpListener)iar.AsyncState).EndGetContext(iar),
    this);
```
to avoid the begin and end delegate allocations on each invocation.